doc/development/pluginsystem.rst

```diff
@@ -0,0 +1,210 @@
+
+* ``base_cls``: The plugin base class (Pydantic BaseModel)
+* ``discriminator_variable``: Field name holding the union (e.g., 'shape', 'plugin')
+* ``discriminator``: Discriminator field name in plugin configs (e.g., 'shape_type', 'plugin_type')
+* ``plugin_registry``: List of registered plugin configuration classes
+* ``plugin_model``: Dynamically generated extensible Pydantic model
+
+PluginSystem Class
+~~~~~~~~~~~~~~~~~~
+
+The central registry stores all plugin families within the class dictionary variable ``_registry`` .
+Each key represents a plugin family name mapping to its corresponding PluginRegistry instance.
+
+Implementation Details
+----------------------
+
+Registration Process
+~~~~~~~~~~~~~~~~~~~~
+
+The decorator-based registration follows a two-step process:
+
+1. **Base Class Registration**: ``@PluginSystem.register()`` creates a registry entry and adds helper methods
+2. **Plugin Registration**: ``@BaseClass.register`` adds plugin classes to the registry and regenerates the union model
+
+Dynamic Model Generation
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The system uses ``pydantic.create_model()`` to dynamically generate extensible models.
+Each registration updates the discriminated union type and recreates the model.
+The discriminator field enables automatic deserialization based on the type identifier.
+
+Usage Examples
+--------------
+
+Basic Plugin Setup
+~~~~~~~~~~~~~~~~~~
+
+The following example demonstrates how to define a plugin base class and register multiple plugin configurations.
+
+.. code-block:: python
+
+    from pydantic import BaseModel
+    from typing import Literal
+    from foamadapter.core.plugin_system import PluginSystem
+
+    @PluginSystem.register(discriminator_variable="shape", discriminator="shape_type")
+    class ShapeInterface(BaseModel):
+        color: str
+        name: str = "default"
+
+    @ShapeInterface.register
+    class CircleConfig(BaseModel):
+        shape_type: Literal["circle"]
+        radius: float
+
+    @ShapeInterface.register
+    class SquareConfig(BaseModel):
+        shape_type: Literal["square"]
+        side: float
+
+Creating and Using Configurations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A usage example creating and using plugin configurations is shown below.
+
+.. code-block:: python
+
+    # Direct instantiation using the plugin model
+    Shape = ShapeInterface.plugin_model
+    circle = Shape(
+        shape={"shape_type": "circle", "radius": 5.0},
+        color="red"
+    )
+
+    # Using the create class method
+    square = ShapeInterface.create(
+        shape={"shape_type": "square", "side": 3.0},
+        color="blue"
+    )
+
+    # Accessing plugin-specific fields
+    assert circle.shape.radius == 5.0
+    assert square.shape.side == 3.0
+
+Runtime Extension
+~~~~~~~~~~~~~~~~~
+
+The plugin system supports runtime registration of new plugin configurations.
+However, this requires rebuilding the Type union which is done by the create method.
```

Comment:
> rebuilding the type union 

Ok so my understanding from this that the type union is basically the entirety of the plugin types of a specific class, e.g. for turbulence it would hould k-epsilon, k-omega as a type union. I think this needs to spelled out explicitly somewhere.